File: Data Structures/Queue/Queue.java
import java.util.Scanner;

class Queue {
    Node head;

    class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    int peek() {
        if (head == null) {
            return -1;
        } else {
            return head.data;
        }
    }

    void enqueue(int data) {
        Node newnode = new Node(data);

        if (head == null) {
            head = newnode;
            return;
        }

        Node ptr = head;
        while (ptr.next != null) {
            ptr = ptr.next;
        }

        ptr.next = newnode;
    }

    boolean dequeue() {
        if (head == null) {
            return false;
        }

        head = head.next;
        return true;
    }

    boolean search(int key) {
        Node ptr = head;
        while (ptr != null) {
            if (ptr.data == key) {
                return true;
            }
            ptr = ptr.next;
        }
        return false;
    }

    void display() {
        Node ptr = head;
        if (ptr==null){
            System.out.println("Queue is empty");
        }
        while (ptr != null) {
            System.out.print(ptr.data + " ");
            ptr = ptr.next;
        }
    }

    boolean isEmpty() {
        return head == null;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        Queue queue = new Queue();

        while (true) {
            System.out.println("\nChoose an operation:");
            System.out.println("1. Enqueue");
            System.out.println("2. Dequeue");
            System.out.println("3. Peek");
            System.out.println("4. Search");
            System.out.println("5. Display");
            System.out.println("6. Check if empty");
            System.out.println("7. Exit");

            int choice = sc.nextInt();

            switch (choice) {
                case 1:
                    System.out.println("Enter data to enqueue: ");
                    int data = sc.nextInt();
                    queue.enqueue(data);
                    break;

                case 2:
                    boolean resp= queue.dequeue();
                    if (resp == false) {
                        System.out.println("Queue is empty");
                    } else {
                        System.out.println("Element dequed");
                    }
                    break;

                case 3:
                    int peekValue = queue.peek();
                    if (peekValue == -1) {
                        System.out.println("Queue is empty");
                    } else {
                        System.out.println("Peek value is " + peekValue);
                    }
                    break;

                case 4:
                    System.out.println("Enter key to search: ");
                    int key = sc.nextInt();
                    boolean isFound = queue.search(key);
                    if (isFound) {
                        System.out.println(key + " is present in the queue");
                    } else {
                        System.out.println(key + " is not present in the queue");
                    }
                    break;

                case 5:
                    System.out.println("Queue elements:");
                    queue.display();
                    break;

                case 6:
                    if (queue.isEmpty()) {
                        System.out.println("Queue is empty");
                    } else {
                        System.out.println("Queue is not empty");
                    }
                    break;

                case 7:
                    System.out.println("Exiting...");
                    System.exit(0);

                default:
                    System.out.println("Invalid choice");
            }
        }
    }
}

File: Data Structures/Queue/Queue.cpp
#include <iostream>

class Node {
public:
    int data;
    Node* next;

    Node(int value) {
        data = value;
        next = nullptr;
    }
};

class Queue {
private:
    Node* front; // pointer to the front element of the queue
    Node* rear;  // pointer to the rear element of the queue

public:
    Queue() {
        front = nullptr;
        rear = nullptr;
    }

    bool isEmpty() {
        return front == nullptr;
    }

    void enqueue(int value) {
        Node* newNode = new Node(value);

        if (isEmpty()) {
            front = newNode;
            rear = newNode;
        } else {
            rear->next = newNode;
            rear = newNode;
        }

        std::cout << value << " enqueued to the queue.\n";
    }

    void dequeue() {
        if (isEmpty()) {
            std::cout << "Queue is empty. Cannot dequeue element.\n";
            return;
        }

        Node* temp = front;
        int value = temp->data;
        front = front->next;

        if (front == nullptr) {
            rear = nullptr;
        }

        delete temp;

        std::cout << value << " dequeued from the queue.\n";
    }

    int peek() {
        if (isEmpty()) {
            std::cout << "Queue is empty. No element to peek.\n";
            return -1;
        }

        return front->data;
    }

    void display() {
        if (isEmpty()) {
            std::cout << "Queue is empty.\n";
            return;
        }

        std::cout << "Queue contents: ";
        Node* temp = front;
        while (temp != nullptr) {
            std::cout << temp->data << " ";
            temp = temp->next;
        }
        std::cout << std::endl;
    }
};

int main() {
    Queue queue;
    int choice, value;

    do {
        // Display menu
        std::cout << "\nQueue Menu\n";
        std::cout << "====================\n";
        std::cout << "1. Enqueue\n";
        std::cout << "2. Dequeue\n";
        std::cout << "3. Peek\n";
        std::cout << "4. Display\n";
        std::cout << "5. Exit\n";
        std::cout << "Enter your choice: ";
        std::cin >> choice;

        switch (choice) {
            case 1:
                std::cout << "Enter value to enqueue: ";
                std::cin >> value;
                queue.enqueue(value);
                break;
            case 2:
                queue.dequeue();
                break;
            case 3:
                std::cout << "Front element of queue: " << queue.peek() << std::endl;
                break;
            case 4:
                queue.display();
                break;
            case 5:
                std::cout << "Exiting program...\n";
                break;
            default:
                std::cout << "Invalid choice. Please try again.\n";
        }
    } while (choice != 5);

    return 0;
}
